bin/kano-updater-recovery

```diff
@@ -0,0 +1,351 @@
+
+
+def _start_systemd_service(service):
+    """
+    Helper function to launch SystemD services
+
+    Args:
+        service (str): Name of the SystemD service to launch
+    """
+
+    import os
+    os.system('systemctl start {}'.format(service))
+
+
+def show_recovery_splash():
+    """
+    Shows the main recovery splash
+    """
+
+    _start_systemd_service('recovery-running-updater-splash')
+    import time
+    time.sleep(0.2)
+    _stop_systemd_service('recovery-boot-splash')
+
+
+def show_success_splash():
+    """
+    Shows the splash for a successful recovery
+
+    Returns:
+        int: Return code. Should be one of the members of the :class:`RC` class
+    """
+
+    _start_systemd_service('recovery-success-splash')
+    import time
+    time.sleep(0.2)
+    _stop_systemd_service('recovery-running-updater-splash')
+
+    return RC.SUCCESS
+
+
+def show_err(err_code):
+    """
+    Shows the error splash screen associated with the error code. Requires a
+    directory in :attr:`kano_updater.ui.paths.MEDIA_PATH`.
+
+    Args:
+        err_code (int): Error code to show. Should be a member of :class:`RC`.
+
+    Returns:
+        int: The supplied return code
+    """
+
+    err_str = RC.rc_to_str(err_code)
+    err_service = 'recovery-err-splash@{}'.format(err_str)
+    _start_systemd_service(err_service)
+
+    import time
+    time.sleep(0.2)
+    _stop_systemd_service('recovery-running-updater-splash')
+
+    return err_code
+
+
+def get_wifi_retries():
+    """
+    Calculates how many times the recovery mechanism has encountered the WiFi
+    connection failure screen.
+
+    Returns:
+        int: WiFi connection failure count
+    """
+
+    try:
+        with open(RECOVERY_STATUS_PATH, 'r') as status:
+            return int(status.read().strip())
+    except (IOError, ValueError):
+        return 0
+
+
+def increment_wifi_retries():
+    """
+    Increase the count of WiFi failures
+    """
+
+    try:
+        retries = get_wifi_retries()
+
+        with open(RECOVERY_STATUS_PATH, 'w') as status:
+            status.write(str(retries + 1))
+    except (IOError, TypeError):
+        pass
+
+
+def reset_wifi_retries():
+    """
+    Resets the count of recovery mechanism has encountered the WiFi
+    connection failure screen.
+    """
+
+    try:
+        import os
+        os.remove(RECOVERY_STATUS_PATH)
+    except (IOError, OSError):
+        pass
+
+
+def is_err_being_shown():
+    """Check if a UFx error splash is currently being shown.
+
+    Returns:
+        bool: Whether a recovery-err-splash service is active
+    """
+
+    import os
+    return (os.system('systemctl is-active -q recovery-err-splash@*.service') == 0)
+
+
+def run_updater():
+    """Launch the updater recovery in the background
+
+    Assumes the state was already checked to be INSTALLING_UPDATES.
+
+    Returns:
+        subprocess.Popen: Launched updater process
+    """
+
+    import os
+    import subprocess
+
+    return subprocess.Popen(
```

Comment:
Can we add a short comment on why we need to `setsid` the new process?

Reply:
This was to assign an id to the entire process group to make stopping it more reliable. Updated the comment.